addons/folo/common/src/main/java/org/commonjava/indy/folo/change/FoloTrackingAdjustListener.java

```diff
@@ -0,0 +1,128 @@
+/**
+ * Copyright (C) 2011-2017 Red Hat, Inc. (https://github.com/Commonjava/indy)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.commonjava.indy.folo.change;
+
+import org.commonjava.indy.data.StoreDataManager;
+import org.commonjava.indy.folo.conf.FoloConfig;
+import org.commonjava.indy.folo.data.FoloRecordCache;
+import org.commonjava.indy.folo.model.TrackedContent;
+import org.commonjava.indy.folo.model.TrackedContentEntry;
+import org.commonjava.indy.folo.model.TrackingKey;
+import org.commonjava.indy.model.core.StoreKey;
+import org.commonjava.indy.promote.change.PromoteChangeManager;
+import org.commonjava.indy.promote.change.event.PathsPromoteCompleteEvent;
+import org.commonjava.indy.promote.change.event.PromoteCompleteEvent;
+import org.commonjava.indy.promote.model.PathsPromoteRequest;
+import org.commonjava.indy.promote.model.PathsPromoteResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.event.Observes;
+import javax.inject.Inject;
+
+import java.util.Set;
+
+@ApplicationScoped
+public class FoloTrackingAdjustListener
+{
+
+    private final Logger logger = LoggerFactory.getLogger( getClass() );
+
+    @Inject
+    private FoloConfig foloConfig;
+
+    @Inject
+    private FoloRecordCache recordManager;
+
+    @Inject
+    private PromoteChangeManager promoteChangeManager;
+
+    @Inject
+    private StoreDataManager storeManager;
+
+    public void onPromoteComplete( @Observes final PromoteCompleteEvent event )
+    {
+        logger.trace( "Promote COMPLETE: {}", event );
+
+        if ( event instanceof PathsPromoteCompleteEvent )
+        {
+            handlePathsPromoteComplete( ( (PathsPromoteCompleteEvent) event ).getPromoteResult() );
+        }
+    }
+
+    private void handlePathsPromoteComplete( PathsPromoteResult promoteResult )
+    {
+        PathsPromoteRequest req = promoteResult.getRequest();
+        StoreKey source = req.getSource();
+        StoreKey target = req.getTarget();
+
+        TrackingKey trackingKey = getTrackingKey( source );
```

Comment:
This is a problem, actually. It works for hosted repositories, when we upload things...but when we download things as a result of a build that uses a tracking key, we won't be able to get that key back from the repository name. For instance, `build_org.commonjava.indy-indy-1.2.3.redhat-1` might download `ch.qos.logback:logback` from `remote:central`. We would still promote things from these remote repos, just to make sure they're available for later build reproduction...which leaves us with lost information, since we'll have no way to retrieve the tracking key(s) using that path.

In fact, that brings up another point: more than one tracking key could reference ch.qos.logback:logback in central. If we promote it into some hosted repo `shared-imports`, we really need to adjust all of those tracking records.

I wonder how heavy it would be to index those tracking records so we could use the ISPN Query interface? I know we've had some trouble with scaling that using filesystem persistence on NFS in the past...however, since we're looking at using something like postgres to persist our ISPN caches for content indexing and such, I wonder if that would make Query a usable feature again?

If we can use Query, we could index by tracking number and path maybe? I'm open to ideas on this one.

Reply:
I thought of only hosted-to-hosted promotion. I reckon there are some trickinesses to do such for remote repositories. As you mentioned, we don't know how many records are affected. We have to do a query for all those and update their downloads entries. If this is not pnc wanted immediately, can we put off this and create a jira to track?

Reply:
Yep, I think we can probably go for the simple solution for now. Longer term, we're probably going to need a larger refactor of the whole Folo concept.